### PR TITLE
[interop][SwiftToCxx] add support for exposing non-isolated actor declarations

### PR DESCRIPTION
**Pull request by hyp**
This exposes Swift actor classes to C++. Only non-isolated declarations get bridged to C++ though.

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

File: test/Interop/SwiftToCxx/class/swift-actor-in-cxx.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  }
+  deinit {
+    print("destroy ActorWithField")
+  }
+
+  public func isolatedMethod() {}
+
+  public nonisolated func method() {
+      print("nonisolated method")
+  }
```
Comment (ktoso) [inline test/Interop/SwiftToCxx/class/swift-actor-in-cxx.swift:30]: I wonder if you'd also want to add a 

```
public nonisolated func methodAsync() async 
```

and confirm it does not import; though I guess this perhaps is already handled -- it's the same as any normal async function after all.

File: test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift [hunk 2]
```diff
@@ -10,9 +10,49 @@
...
+    public func unsupported() {}
+
+    @_expose(Cxx) // expected-error {{actor-isolated property 'prop' can not be exposed to C++}}
+    public var prop: Int { 42 }
+
+    @_expose(Cxx) // expected-error {{actor-isolated distributed instance method 'dist()' can not be exposed to C++}}
+    distributed public func dist() {}
+
+    @_expose(Cxx) // ok
+    public init() {
```
Comment (ktoso) [inline test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift:50]:
```suggestion
    public init(actorSystem: LocalTestingDistributedActorSystem) {
        self.actorSystem = actorSystem
```

is what you'd need to do to make sure this is a 100% correct actor; this should have failed to compile tbh as it is declared now - but the other errors probably prevented it from going into those checks

File: test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift [hunk 3]
```diff
@@ -10,9 +10,49 @@
...
+
+    @_expose(Cxx) // ok
+    nonisolated public func supported() {}
+
+    @_expose(Cxx) // ok
+    nonisolated public var prop2: Int { 42 }
+}
+
+@_expose(Cxx)
+public distributed actor DistributedActorClass {
```
Comment (ktoso) [inline test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift:37]: Hooray! Semantics look good 👍

File: lib/AST/SwiftNameTranslation.cpp [hunk 4]
```diff
@@ -167,11 +167,8 @@ swift::cxx_translation::getNameForCxx(const ValueDecl *VD,
 
 swift::cxx_translation::DeclRepresentation
 swift::cxx_translation::getDeclRepresentation(const ValueDecl *VD) {
-  if (auto *CD = dyn_cast<ClassDecl>(VD)) {
-    // Actors are not exposable to C++.
-    if (CD->isAnyActor())
-      return {Unsupported, UnrepresentableActorClass};
-  }
+  if (getActorIsolation(const_cast<ValueDecl *>(VD)).isActorIsolated())
+    return {Unsupported, UnrepresentableIsolatedInActor};
```
Comment (ktoso) [inline lib/AST/SwiftNameTranslation.cpp:171]: awesome, glad that this indeed covered all the bases :-)

**Review by ktoso — APPROVED** [review]
LGTM! Might want to add a test with `nonisolated + async` method too though

File: test/Interop/SwiftToCxx/class/swift-actor-in-cxx.swift (hunk 1 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/class/swift-actor-in-cxx.swift:30]: async functions are not bridged already, we cover that. But I can add as a follow-up!

File: test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift (hunk 2 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/unsupported/unsupported-decls-in-cxx.swift:50]: Sounds good, I'll commit this as a follow-up!

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
source compatibility suite failures are unrelated, will merge:

```
Failures:
  FAIL: CoreStore, 4.2, 286360, CoreStore watchOS, generic/platform=watchOS
  FAIL: CoreStore, 4.2, 286360, CoreStore tvOS, generic/platform=tvOS
  FAIL: CoreStore, 4.2, 286360, CoreStore OSX, generic/platform=macOS
  FAIL: CoreStore, 4.2, 286360, CoreStore iOS, generic/platform=iOS
  FAIL: CoreStore, 4.0, 83e608, CoreStore watchOS, generic/platform=watchOS
  FAIL: CoreStore, 4.0, 83e608, CoreStore tvOS, generic/platform=tvOS
  FAIL: CoreStore, 4.0, 83e608, CoreStore OSX, generic/platform=macOS
  FAIL: CoreStore, 4.0, 83e608, CoreStore iOS, generic/platform=iOS
========================================
```

**Pull request merged**